File: DotNetCorePracticalView/CS.NodeJS.LoadTest.WebApp/webapp/data/pgsql/dataMigrationPgsqlDAO.js
let pg = require('pg');
let schema = require('../loadTestSchema.js');
let appSettings = require('../../config/appsettings.js');

module.exports = {

	createAsync: (pgsqlConfig) => {

		let dao = {
			_isConnOpen: false,
			_currentDatabase: '',
			_dbConfig: pgsqlConfig ? pgsqlConfig : appSettings.databaseConfig.pgsql,
			_client: null,
			_retryTimeout: null
		};

		dao._closeConnection = () => {

			dao._client.end();
			dao._isConnOpen = false;
			dao._currentDatabase = '';
			dao._client = null;
		};

		dao._openConnectionAsync = (config) => {

			let retryCount = 0;
			let retryTimeout = null;


			if (!dao._client) {
				dao._client = new pg.Client(config);
			}

			return new Promise((res, rej) => {

				let connectCallback = (err) => {

					if (err) {

						if (retryCount == 2) {
							rej(err);
							return;
						}

						retryCount++;
						console.warn('DataMigration connection error. try:' + retryCount + ' starting retry.')

						retryTimeout = setTimeout(() => {

							dao._closeConnection();
							dao._client = new pg.Client(config);
							dao._client.connect(connectCallback);
						}, 10000);

						return;
					}

					if (retryCount > 0) {
						console.log("DataMigration connection success");
					}

					if (retryTimeout) {
						clearTimeout(retryTimeout);
					}

					dao._isConnOpen = true;
					dao._currentDatabase = config.database;
					res();
				};

				dao._client.connect(connectCallback);
			});
		};

		dao._changeDatabaseAsync = (db) => {

			let config = {
				user: dao._dbConfig.rootConn.user,
				password: dao._dbConfig.rootConn.password,
				host: dao._dbConfig.rootConn.host,
				port: dao._dbConfig.rootConn.port,
				database: db
			};

			if (!dao._isConnOpen) {
				return dao._openConnectionAsync(config);
			}

			//change database if needed
			if (dao._currentDatabase != config.database) {

				dao._closeConnection();
				return dao._openConnectionAsync(config);
			}

			return new Promise((res, rej) => { res(); });
		};

		dao._createDatabaseAsync = () => {

			//check if app database exists
			let selectDbCmd = {
				text: 'select 1 as "db" from pg_database where datname = $1',
				values: [dao._dbConfig.conn.database]
			};

			let createDbCmd = {};
			createDbCmd.text = 'create database "' + dao._dbConfig.conn.database + '"';

			createDbCmd.text += ' with template=\'template0\' encoding=\'UTF8\' tablespace=pg_default';
			createDbCmd.text += ' lc_collate=\'' + dao._dbConfig.collate + '\'';

			createDbCmd.text += ' lc_ctype=\'' + dao._dbConfig.ctype + '\' connection limit= -1; '
			createDbCmd.values = [];

			return new Promise((res, rej) => {

				dao._client.query(selectDbCmd)
					.then((result) => {

						if (result.rows.length == 0) {

							dao._client.query(createDbCmd)
								.then((result_2) => { res(); })
								.catch(rej)

							return;
						}
						res();
					})
					.catch(rej);
			});
		};

		dao._createSchemaVersionAsync = () => {

			let cmd = {};
			cmd.text = 'create table if not exists "' + schema.schemaVersion.table + '"';
			cmd.text += ' ("' + schema.schemaVersion.versionColumn + '" char(5) not null);';

			return new Promise((res, rej) => {

				dao._client.query(cmd)
					.then(() => { res(); })
					.catch(rej);
			});
		};

		dao._validateMigrationOperation = (operation, fromVersion, toVersion) => {

			if (fromVersion || toVersion != dao.getLatestVersion()) {
				throw new Error(operation + ' from version "' + fromVersion + '"' + ' to version "' + toVersion + '" not implemented.');
			}

			if (fromVersion == toVersion) {
				console.log(operation + " - Updated");
				return false;
			}

			return true;
		};

		dao.getCurrentVersionAsync = () => {

			return new Promise((res, rej) => {

				let query = {};
				query.text = 'select "' + schema.schemaVersion.versionColumn + '"';
				query.text += ' from "' + schema.schemaVersion.table + '"';

				dao._client.query(query)
					.then((result) => {

						if (result.rows && result.rows[0]) {
							res(result.rows[0][schema.schemaVersion.versionColumn]);
						}
						else {
							res(null);
						}
					})
					.catch(rej);
			});
		};

		dao.migrateDbmsUsersAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate DBMS Users', fromVersion, toVersion)) {
				return new Promise((res, rej) => { res(); });
			}

			//create user
			let cmd = {};
			cmd.text = 'create user ' + dao._dbConfig.conn.user;
			cmd.text += ' with encrypted password \'' + dao._dbConfig.conn.password + '\';';

			//grant access to connect
			cmd.text += ' grant connect, temporary on database ' + dao._dbConfig.conn.database
			cmd.text += ' TO ' + dao._dbConfig.conn.user + ';';

			//grant access to tables
			cmd.text += ' grant select, insert, update, delete, truncate, references, trigger';
			cmd.text += ' on all tables in schema public to ' + dao._dbConfig.conn.user + ';';

			return new Promise((res, rej) => {

				dao._client.query(cmd)
					.then(() => { res(); })
					.catch(rej);
			});
		};

		dao.migrateIdentityAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate Identity', fromVersion, toVersion)) {
				return new Promise((res, rej) => { res(); });
			}

			//create table identity
			let cmd = {}
			cmd.text = 'create table "' + schema.identity.table + '"';
			cmd.text += ' ("' + schema.identity.identityNameColumn + '" varchar(100) not null';
			cmd.text += ' constraint ' + schema.identity.IdentityNameIndex + ' primary key,';
			cmd.text += ' "' + schema.identity.dataColumn + '" json not null);';

			return new Promise((res, rej) => {

				dao._client.query(cmd)
					.then(() => { res(); })
					.catch(rej);
			});
		};

		dao.migrateSchemaVersionAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate Schema Version', fromVersion, toVersion)) {
				return new Promise((res, rej) => { });
			}

			//update version
			let cmd = {};
			cmd.text = 'update "' + schema.schemaVersion.table + '"';

			cmd.text += ' set "' + schema.schemaVersion.versionColumn + '" = $1;';
			cmd.values = [toVersion];

			return new Promise((res, rej) => {

				dao._client.query(cmd)
					.then((result) => {

						//if there is no record insert one
						if (result.rowCount == 0) {

							cmd.text = 'insert into "' + schema.schemaVersion.table + '"'
							cmd.text += ' ("' + schema.schemaVersion.versionColumn + '") values($1);'

							dao._client.query(cmd)
								.then(() => { res(); })
								.catch(rej);
							return;
						}
						res();
					})
					.catch(rej);
			});
		};

		dao.disposeAsync = () => {

			dao._closeConnection();
			return new Promise((res, rej) => { res(); });
		};

		dao.getLatestVersion = () => { return '1.0.0'; }

		dao.migrateTestResultAsync = (fromVersion, toVersion) => {

			if (!dao._validateMigrationOperation('Migrate Test Result', fromVersion, toVersion)) {
				return new Promise((res, rej) => { res(); });
			}

			//create table TestResult
			let cmd = {};
			cmd.text = 'create table "' + schema.testResult.table + '"';
			cmd.text += ' ("' + schema.testResult.testIdColumn + '" varchar(10) not null,';
			cmd.text += ' "' + schema.testResult.languageColumn + '" varchar(50) null,';
			cmd.text += ' "' + schema.testResult.dataColumn + '" json not null,';

			cmd.text += ' constraint "' + schema.testResult.testResultTestIndex + '"';
			cmd.text += ' unique ("' + schema.testResult.testIdColumn + '","' + schema.testResult.languageColumn + '"));';

			return new Promise((res, rej) => {

				dao._client.query(cmd)
					.then(() => { res(); })
					.catch(rej);
			});
		};

		return new Promise((res, rej) => {

			dao._openConnectionAsync(dao._dbConfig.rootConn)
				.then(() => {

					dao._createDatabaseAsync()
						.then(() => {

							dao._changeDatabaseAsync(dao._dbConfig.conn.database)
								.then(() => {

									dao._createSchemaVersionAsync()
										.then(() => { res(dao); })
										.catch(rej);
								})
								.catch(rej);
						})
						.catch(rej);
				})
				.catch(rej);
		});
	}
}

File: CS.DotNetCore.LoadTest/src/CS.DotNetCore.LoadTest.WebApp/Data/Pgsql/DataMigrationPgsqlDAO.cs
namespace CS.DotNetCore.LoadTest.WebApp.Data.Pgsql
{
    using Config;
    using Npgsql;
    using NpgsqlTypes;
    using Schema;
    using System.Data;
    using System.Text;
    using Util.Data;
    using System;

    internal class DataMigrationPgsqlDAO : BaseDataMigrationDAO, IDataMigrationDAO
    {
        private IDatabaseConfig _config;
        private NpgsqlConnection _connection;

        public override string LatestVersion
        {
            get
            {
                return "1.0.0";
            }
        }

        internal DataMigrationPgsqlDAO(IDatabaseConfig config)
        {
            _config = config;
            _connection = new NpgsqlConnection(config.AdminString);
            OpenConnection(_connection.Open);

            CreateDatabase();
            ChangeDatabase(config.DataBase);

            CreateSchemaVersion();
        }

        private void ChangeDatabase(string database = "postgres")
        {
            if (_connection.Database != database)
            {
                _connection.Close();
                _connection.Dispose();

                var connString = new NpgsqlConnectionStringBuilder(_config.AdminString);
                connString.Database = database;

                _connection = new NpgsqlConnection(connString.ConnectionString);
                _connection.Open();
            }
        }

        private void CreateDatabase()
        {
            var sql = new StringBuilder();
            sql.Append("select 1 from pg_database where datname = :dbname");

            var parameters = new NpgsqlParameter[1]
            {
                new NpgsqlParameter("dbName", NpgsqlDbType.Varchar)
                {
                    Direction = ParameterDirection.Input,
                    IsNullable = false,
                    NpgsqlValue = _config.DataBase,
                }
            };

            var dbExists = DataHelper.ExecuteScalar(_connection, sql.ToString(), parameters);

            //create database if not exists
            if (dbExists == null)
            {
                sql.Clear();

                sql.AppendFormat("create database \"{0}\"", _config.DataBase);
                sql.Append(" with template='template0' encoding='UTF8' tablespace=pg_default");

                sql.AppendFormat(" lc_collate='{0}' lc_ctype='{1}' connection limit=-1;", _config.Collate, _config.Ctype);
                DataHelper.ExecuteNonQuery(_connection, sql.ToString());
            }
        }

        private void CreateSchemaVersion()
        {
            var sql = new StringBuilder();
            sql.AppendFormat("create table if not exists \"{0}\"", SchemaVersionSchema.Table);

            sql.AppendFormat(" (\"{0}\" char(5) not null);", SchemaVersionSchema.VersionColumn);
            DataHelper.ExecuteNonQuery(_connection, sql.ToString());
        }

        public string GetCurrentVersion()
        {
            //resolve current version
            var sql = string.Format("select \"{0}\" from \"{1}\";",
                SchemaVersionSchema.VersionColumn, SchemaVersionSchema.Table);

            var version = DataHelper.ExecuteScalar(_connection, sql.ToString());
            return version == null ? null : version.ToString();
        }

        public void MigrateDbmsUsers(string fromVersion, string toVersion)
        {
            if (!ValidateMigrationOperation("Migrate DBMS Users", fromVersion, toVersion))
            {
                return;
            }

            var sql = new StringBuilder();

            //create user
            sql.AppendFormat("create user {0} with encrypted password '{1}';",
                _config.User, _config.Password);

            //grant access to connect on database
            sql.AppendFormat(" grant connect, temporary on database {0} TO {1};",
                _config.DataBase, _config.User);

            //grant access to tables
            sql.AppendFormat(" grant select, insert, update, delete, truncate, references, trigger on all tables in schema public to {0};",
                _config.User);

            DataHelper.ExecuteNonQuery(_connection, sql.ToString());
        }
        
        public void MigrateIdentity(string fromVersion, string toVersion)
        {
            if (!ValidateMigrationOperation("Migrate Identity", fromVersion, toVersion))
            {
                return;
            }

            var sql = new StringBuilder();

            sql.AppendFormat("create table \"{0}\"(\"{1}\" varchar(100) not null constraint {2} primary key,",
                IdentitySchema.Table, IdentitySchema.IdentityNameColumn, IdentitySchema.IdentityNameIndex);

            sql.AppendFormat(" \"{0}\" json not null);", IdentitySchema.DataColumn);
            DataHelper.ExecuteNonQuery(_connection, sql.ToString());
        }

        public void MigrateSchemaVersion(string fromVersion, string toVersion)
        {
            if (!ValidateMigrationOperation("Migrate Schema Version", fromVersion, toVersion))
            {
                return;
            }

            var parameter = new NpgsqlParameter[1]
            {
                new NpgsqlParameter(SchemaVersionSchema.VersionColumn, NpgsqlDbType.Char)
                {
                    Direction = ParameterDirection.Input,
                    IsNullable = false,
                    NpgsqlValue = toVersion
                }
            };

            var sql = string.Format("update \"{0}\" set \"{1}\"=:{1};", SchemaVersionSchema.Table, SchemaVersionSchema.VersionColumn);
            var affected = DataHelper.ExecuteNonQuery(_connection, sql, parameter);

            if (affected == 0)
            {
                sql = string.Format("insert into \"{0}\" (\"{1}\") values (:{1});",
                    SchemaVersionSchema.Table, SchemaVersionSchema.VersionColumn);

                parameter = new NpgsqlParameter[1] { parameter[0].Clone() };
                DataHelper.ExecuteNonQuery(_connection, sql, parameter);
            }
        }

        public void MigrateTestResult(string fromVersion, string toVersion)
        {
            if (!ValidateMigrationOperation("Migrate Test Result", fromVersion, toVersion))
            {
                return;
            }

            var sql = new StringBuilder();
            sql.AppendFormat("create table \"{0}\"", TestResultSchema.Table);

            sql.AppendFormat(" (\"{0}\" varchar(10) not null,", TestResultSchema.TestIdColumn);
            sql.AppendFormat(" \"{0}\" varchar(50) null,", TestResultSchema.LanguageColumn);
            sql.AppendFormat(" \"{0}\" json not null,", TestResultSchema.DataColumn);

            sql.AppendFormat(" constraint \"{0}\" unique (\"{1}\",\"{2}\"));",
                TestResultSchema.TestResultTestIndex, TestResultSchema.TestIdColumn, TestResultSchema.LanguageColumn);

            DataHelper.ExecuteNonQuery(_connection, sql.ToString());
        }

        public void Dispose()
        {
            _config = null;
            _connection.Close();

            _connection.Dispose();
            _connection = null;
        }
    }
}
